### PR TITLE
Fix zero fee check to show free text

### LINKED ISSUE #2015: [1.30]Show 'Free' instead of 0 when no fees for an order
Related to #1992
1. Using https://www.impersonator.xyz/ connect to this account 0xDe1E8EedD42CD284a3052d987502d0B24c633FEE (this is a liquidity provider)
2. Open Swap page and specify a token pair, amount to trade
3. Check the Fee field

**AR**: fee show 0

**ER**: fee shows '🎉 Free!'
For ETH orders: show '🎉 Free!'+Gas' and it would be nice to remove USD estimation
image

### PR DESCRIPTION
**Pull request by tukantje**
# Summary

Fixes #2015 

Our implementation for checking whether or not there is a zero / non-zero fee was incorrect as it was looking for falsy strings. Also, showing fiat equivalent had no checks regarding whether it should be shown, as when it is free, it is not necessary.

I've added a utility to check for a string amount being a valid non-zero amount, and updated the logic to show free correctly as shown down below.

Screenshot 2023-02-17 at 03 37 49
Screenshot 2023-02-17 at 03 37 28

  # To Test

Free case:
1. Using https://www.impersonator.xyz/ connect to this account 0xDe1E8EedD42CD284a3052d987502d0B24c633FEE (this is a liquidity provider)
2. Open Swap page and specify a token pair, amount to trade
3. Check the Fee field
4. Expect;
- [] Free should be shown without any numbers.

Non-free case:
1. Using any non liquidity provider wallet connect to your account
2. Open swap page and specify a token pair that would have fees
3. Check the fee field
4. Expect;
- [] Fee should be shown with its fiat equivalent next to it.

### CODE REVIEW

**Review by shoom3301 — APPROVED** [review]

**Review by elena-zh — CHANGES_REQUESTED** [review]
Hey @tukantje , good job!
However, some issues from my side: 
1. I see 'Free' for trades where fee is not 0:
 - Gorli: WETH-DAI token pair
gorli
 - Gnosis chain: WXDAI-WBTC
gnosis chain
2. Besides, for some cases it shows 'Free' for ETH/XDAI orders, hover, should show Fee+Gas value
fee+gas
3. For a liquidity provider it would be great to show 'Free!+gas' when place a ETH/XDAI order, and show 'Free' in the Recieve (incl.fee) tooltip
LP


Could you please fix all these issues?

**Comment by tukantje** [comment]
@elena-zh of course - investigating.

**Review by shoom3301 — APPROVED** [review]

File: src/cow-react/modules/swap/containers/Row/RowFee/index.tsx [hunk 1]
```diff
@@ -71,14 +81,16 @@ export function RowFee({ trade, fee, feeFiatValue, allowsOffchainSigning, showHe
     const feeAmountWithCurrency = `${smartFeeTokenValue} ${formatSymbol(feeCurrencySymbol)} ${
       isEthFLow ? ' + gas' : ''
     }`
-    const feeToken = smartFeeTokenValue ? feeAmountWithCurrency : '🎉 Free!'
+    console.log(smartFeeTokenValue, smartFeeFiatValue)
+    const feeToken = isValidNonZeroAmount(smartFeeTokenValue) ? feeAmountWithCurrency : '🎉 Free!'
```
Comment (anxolin) [inline src/cow-react/modules/swap/containers/Row/RowFee/index.tsx:85]: this is not part of this PR, so no need to address it, but for me is a red-flag to see a `feeToken` var that might be an amount or a label.

We are in a smart component, I believe the responsibility should be to get the value of the fee and pass it down. 
It should end up arriving to a dumb component, which will have the logic of 

if (noFee) {
  // Show is free label
} else {
   // show fee
}

Comment (tukantje) [reply src/cow-react/modules/swap/containers/Row/RowFee/index.tsx:85]: I think a small refactor could be nice here, I agree wholeheartedly with this. Perhaps with a hook that contains logic & a pure component?

**Comment by elena-zh** [comment]
> Hey @tukantje , good job! However, some issues from my side:
> 
> 1. I see 'Free' for trades where fee is not 0:
> 
> * Gorli: WETH-DAI token pair
>   gorli
> * Gnosis chain: WXDAI-WBTC
>   gnosis chain
> 
> 2. Besides, for some cases it shows 'Free' for ETH/XDAI orders, hover, should show Fee+Gas value
>    fee+gas
> 3. For a liquidity provider it would be great to show 'Free!+gas' when place a ETH/XDAI order, and show 'Free' in the Recieve (incl.fee) tooltip
>    LP
> 
> Could you please fix all these issues?

Hey @tukantje , cases 1-2 are no longer reproducible. 
Only the 3rd case is left to fix here. 

Thanks

**Comment by anxolin** [comment]
@tukantje should we push this one through the finish line?

File: src/cow-react/modules/swap/containers/Row/RowFee/index.tsx [hunk 2]
```diff
@@ -71,14 +81,16 @@ export function RowFee({ trade, fee, feeFiatValue, allowsOffchainSigning, showHe
     const feeAmountWithCurrency = `${smartFeeTokenValue} ${formatSymbol(feeCurrencySymbol)} ${
       isEthFLow ? ' + gas' : ''
     }`
-    const feeToken = smartFeeTokenValue ? feeAmountWithCurrency : '🎉 Free!'
+    console.log(smartFeeTokenValue, smartFeeFiatValue)
```
Comment (shoom3301) [inline src/cow-react/modules/swap/containers/Row/RowFee/index.tsx:84]: I think the console.log should be removed

**Review by shoom3301 — APPROVED** [review]

File: src/cow-react/modules/swap/containers/Row/RowFee/index.tsx (hunk 2 above)
Comment (tukantje) [reply src/cow-react/modules/swap/containers/Row/RowFee/index.tsx:84]: Great point - removing it

**Comment by tukantje** [comment]
Cherry picked `7e70f6b0d885f4cbdc32177d4d8b8bd9e77279f0` due to deployments having wrong infura instance.

**Comment by tukantje** [comment]
> @tukantje should we push this one through the finish line?



> @tukantje should we push this one through the finish line?



> > Hey @tukantje , good job! However, some issues from my side:
> > 
> > 1. I see 'Free' for trades where fee is not 0:
> > 
> > 
> > * Gorli: WETH-DAI token pair
> >   gorli
> > * Gnosis chain: WXDAI-WBTC
> >   gnosis chain
> > 
> > 
> > 2. Besides, for some cases it shows 'Free' for ETH/XDAI orders, hover, should show Fee+Gas value
> >    fee+gas
> > 3. For a liquidity provider it would be great to show 'Free!+gas' when place a ETH/XDAI order, and show 'Free' in the Recieve (incl.fee) tooltip
> >    LP
> > 
> > Could you please fix all these issues?
> 
> Hey @tukantje , cases 1-2 are no longer reproducible. Only the 3rd case is left to fix here.
> 
> Thanks



> @tukantje should we push this one through the finish line?

I've addressed the last case as well;
Screenshot 2023-04-11 at 12 03 35

**Comment by elena-zh** [comment]
Hey @tukantje , great job!
The last case that is still not fixed is to change '+0' and '-0' to 'Free!'  in the receive/from tooltips when connected to a liquidity provider account (0xDe1E8EedD42CD284a3052d987502d0B24c633FEE). Could you please take a look at it? 
free2
Screenshot_2

**Comment by elena-zh** [comment]
On the order confirm modal -0 and +0 is still displayed instead of Free
free
freeee

**Comment by tukantje** [comment]
> On the order confirm modal -0 and +0 is still displayed instead of Free free freeee

Quite a whack a mole 😅 Apparently we have duplicate functionality regarding this, with a todo left for refactor.

I've updated the component to rely on the other one (apparently `FeeInformationTooltip` is a component that shows a row and a tooltip, as opposed to `ReceivedAmountInfoTooltip` which only is the content of a tooltip.).

This should do the trick (I sincerely hope). However, I think we also need another code review on this, as it constitutes a refactor.

**Review by alfetopito — APPROVED** [review]
Nice refactor!

Only one comment regarding the infura keys

File: .env [hunk 3]
```diff
@@ -15,7 +15,7 @@ REACT_APP_DATA_HASH_DEVELOPMENT=0x828569F802B7F8957F76996BDD875674821E41A688541A
 REACT_APP_DATA_HASH_LOCAL=0xA5DAE7A114F1BD6BB9B3FF976150380A95CB18856212DB555C25EF9D7801E9A4
 
 # INFURA KEY
-REACT_APP_INFURA_KEY=1221fd11e90849509afafd330ec7acc6
+REACT_APP_INFURA_KEY=586e7e6b7c7e437aa41f5da496a749f5
```
[the remark this answers is not in the document]
Comment (tukantje) [reply .env:18]: This is quite an old branch, so had to cherry pick it

[the remark this answers is not in the document]
Comment (anxolin) [reply .env:18]: mm, if you have an updated develop and you rebased or merged it, it should be here, no need to cherrypick

**Review by anxolin — COMMENTED** [review]
Wow, so many changes for some apparently innocent small change.

**Comment by tukantje** [comment]
@elena-zh updated this part as well, my mistake.

Screenshot 2023-04-13 at 01 42 29

@anxolin took me a bit to realise, but I apparently had a divergent develop on my local for some reason. Now it should be cleaned up.

File: .env (hunk 3 above)
[the remark this answers is not in the document]
Comment (tukantje) [reply .env]: Good point - had a divergent branch. Fixed it.

**Review by elena-zh — COMMENTED** [review]
LGTM now!

**Pull request merged**